provider/pkg/schema/gen_tags.go

```diff
@@ -0,0 +1,35 @@
+package schema
+
+import (
+	jsschema "github.com/lestrrat-go/jsschema"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+type TagsStyle string
+
+const (
+	TagsStyleUnknown       TagsStyle = ""
+	TagsStyleUntyped       TagsStyle = "untyped"
+	TagsStyleStringMap     TagsStyle = "stringMap"
+	TagsStyleKeyValueArray TagsStyle = "keyValueArray"
+)
+
+func (ctx *context) GetTagsStyle(typeSpec *pschema.TypeSpec, originalSpec *jsschema.Schema) TagsStyle {
+	if typeSpec == nil || originalSpec == nil {
+		return TagsStyleUnknown
+	}
+	// Check for "Any" ref
+	if typeSpec.Ref == "pulumi.json#/Any" {
+		return TagsStyleUntyped
+	}
+	switch typeSpec.Type {
+	case "object":
```

Comment:
[nit] Could simplify with `case "object" && typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type == "string"` and otherwise fall through to `default`. Unless you structured it that way to add more cases later.

Reply:
Yeah, I ended up scrapping the switch as the type name is actually not a clean way of detecting each case.